### DISCUSSION TITLE
Questions & Contributing

### DISCUSSION THREAD

**Discussion by MithrandirCraft** [GriefPrevention/GriefPrevention#1581] [Q&A] [answered]
Browsing through GP's code, some questions came to mind:

I'm surprised you considered preloading the entire config into memory when the plugin starts, why is that?

Why is the main class so cluttered? Why is the entirety of the commands treated in GriefPrevention.java, and why don't you subdivide each command into it's own execution class?

I also suspect your onCommand() method could be optimized if you made a tree-like branching for it, by dividing commands depending on their number of arguments, thus reducing the number of "else" checks.

I suspect if a command doesn't exist, it has GP iterates through all of it's command posibilities. While you could make it so only a smaller group of the commands are iterated, depending on the number of arguments.

You would also avoid having to check if the number of arguments is ok within each command, since that check would already be done right at the beginning (you are currently sometimes checking arg amounts even more than once per command).

I also think you forgot to include some "else"'s right at the beginning. Instead, a ton of if's are being checked one after the other, no matter if one already entered.

I could do this small optimization for you through PR on 1.0, but I'm asuming you guys are already improving the whole plugin on GP 2.0, so IDK if I should even attempt this. You may also have your specific reasons for doing things like this.

Also, the KickFromClaim addon is kinda dead, and I'm wondering if it would be better to actually code a kick command on the main plugin. Still, since u'r working on 2.0, idk if:
- It's even worth it since you're working on an entirely new version.
- I should just attempt to revive KFC, even though it's not open source, or make a new one.
- You guys have already implemented some proper kick commands on 2.0.

In general if you ask me GP is vital for many survival servers, but on mine, it's always the main cause for a plethora shortcomings due to it's design, and I'd like to aliviate at least a few things if possible, even if I'm not that much of an expert in Java. Just tell me how to do that without efforts being discarded on 2.0.

**Comment by Jikoo** [comment, marked as answer]
Going to drop these relevant discussions at the top because this is gonna be a long reply: #1162 for the roadmap and #63 for the v20 discussion thread.

> I'm surprised you considered preloading the entire config into memory when the plugin starts, why is that?

That's basically how all configurations work. If you use the Bukkit configuration, you just don't see it. It's not like you'd read from disk every time, that'd be horrendously slow. If anything, GP's system is more efficient because it doesn't keep around the original YamlConfiguration it loaded from, only the fields it cares about.

>Why is the main class so cluttered? Why is the entirety of the commands treated in GriefPrevention.java, and why don't you subdivide each command into it's own execution class?

I mean, to be brutally honest about it, because it's poorly written. The length of the main class itself is because GP falls into the classic trap of making its main class a [god object](https://en.wikipedia.org/wiki/God_object). It's bad practice, but dividing it all up at once would be too large a diff to be easily readable or tested.

Moving each command to its own subclass is a massive undertaking that brings up a ton of design choices. I.e. is it appropriate to use a third party lib a la ACF? Personally Bukkit's framework is so lacking and GP has so many commands that I'd say yes, but that then inflates the final file, which Robo is against. Designing a new command framework from scratch specifically for GP is a massive pain in the butt, but Bukkit's own system is also pretty simplistic and therefore slow and annoying to program commands for. The only thing Bukkit has going for it is that commands are written for it and can be copied to a new place instantly, but the commands also contain quite a bit of duplicate code and logic that would ideally be removed. If the commands are being rewritten anyway, why not rewrite them to be actually good? It's a bit of a mess.

This all is part of the goal for v17, so contributions on that would be welcome. Make them small if you can, i.e. rather than split out every command at once split out only /accesstrust, /containertrust, etc. that logically should share an executor - many small PRs are much easier to review than one large one!

> I also suspect your onCommand() method could be optimized if you made a tree-like branching for it, by dividing commands depending on their number of arguments, thus reducing the number of "else" checks.

This does not allow you to create particular responses, i.e. "please specify a player" for missing player args. It's an okay idea, but it would be better to just register individual commands individually.

>I suspect if a command doesn't exist, it has GP iterates through all of it's command posibilities. While you could make it so only a smaller group of the commands are iterated, depending on the number of arguments.

If a command doesn't exist in GP it doesn't get handed to GP, that's how the command system works. Since all of the commands are there in that one absolutely gross method, it will handle it... eventually. Bukkit only sends the command to the handler to which it is registered. Since GP doesn't register any command handlers, they all get sent to the main plugin. Bukkit will not send GP nonexistent commands.

>I also think you forgot to include some "else"'s right at the beginning. Instead, a ton of if's are being checked one after the other, no matter if one already entered.

Those return, which completes execution:

```Java
//abandonclaim
if (cmd.getName().equalsIgnoreCase("abandonclaim") && player != null)
{
    return this.abandonClaimHandler(player, false);
}

//abandontoplevelclaim
if (cmd.getName().equalsIgnoreCase("abandontoplevelclaim") && player != null)
{
    return this.abandonClaimHandler(player, true);
}
```

An else statement for either of these would be functionally identical and is unnecessary.


>I could do this small optimization for you through PR on 1.0, but I'm asuming you guys are already improving the whole plugin on GP 2.0, so IDK if I should even attempt this. You may also have your specific reasons for doing things like this.

The v20 from scratch rewrite involved a prohibitive amount of effort beforehand and eventually got scrapped. There won't be a hard rewrite, but most legacy code (i.e. the onCommand and other massive methods, static abuse, god object status, etc.) will eventually ideally vanish. The problem is time, it takes a lot of time. Contributions are welcome. Heck, that's how I ended up here, I'm just some random person who liked GP enough to spend a lot of time on issue management.

>Also, the KickFromClaim addon is kinda dead, and I'm wondering if it would be better to actually code a kick command on the main plugin.

Anything that is currently an addon is unlikely to be added, Robo would vastly prefer to pare down features and pass them to addons to reduce bloat. GP doesn't do a good job (or any job really) of modularity right now, so adding a new feature generally just adds more spaghetti. You could probably just piggyback GP's `/trapped` system to eject players, though I do think that gives up after a while if it can't find a safe spot, which you probably don't want for a kick plugin.

**Comment by MithrandirCraft** [comment]
Thank you for the detialed response. It does clear some stuff up.

I think I'll just put together a "kfc hack" in the meanwhile, and wait for v20 in order to do anything more proper.

If I find any small thing I can contribute to (specially update stuff that gets forgotten or unattended) I will.
